File: sdk/akg-ts/test/indexes.test.ts
// PERF-1: the secondary indexes back listNodesByTag / outboundEdges /
// inboundEdges with O(matches) lookups instead of O(total) full scans, and stay
// consistent across every mutation path (replace, delete, cascade, WAL replay).
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mkdtempSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { open, Store } from '../src/store.js';

interface InternalState {
  tagIndex: Map<string, Set<string>>;
  outIndex: Map<string, Set<string>>;
  inIndex: Map<string, Set<string>>;
}

// `private state` is a TypeScript compile-time modifier, present at runtime —
// reach it for white-box index assertions (mirrors the akg-go in-package test).
function internalState(s: Store): InternalState {
  return (s as unknown as { state: InternalState }).state;
}

const nk = (type: string, id: string) => `${type}\0${id}`;

let dir: string;
beforeEach(() => { dir = mkdtempSync(join(tmpdir(), 'akg-ts-idx-')); });
afterEach(() => { rmSync(dir, { recursive: true }); });

describe('secondary indexes (PERF-1)', () => {
  it('are sized to matches, not total store size', async () => {
    const s = await open(join(dir, 'idx.akg'));
    // A mass of untagged nodes the indexes must NOT have to scan. Kept under the
    // 1000-entry WAL auto-flush threshold so the test exercises indexes, not
    // repeated whole-file rewrites (IO-1's concern).
    const bulk = 500;
    for (let i = 0; i < bulk; i++) s.putNode('bulk', `b${i}`, { title: 'b' }, []);
    for (let i = 0; i < 5; i++) s.putNode('doc', `d${i}`, { title: 'd' }, ['special']);
    s.putNode('hub', 'h', { title: 'hub' }, []);
    for (let i = 0; i < 10; i++) {
      s.putEdge({ type: 'hub', id: 'h' }, 'links', { type: 'bulk', id: `b${i}` }, {});
    }

    const state = internalState(s);
    expect(state.tagIndex.get('special')?.size, 'tag index must hold only the 5 matches').toBe(5);
    expect(state.outIndex.get(nk('hub', 'h'))?.size).toBe(10);
    expect(state.inIndex.get(nk('bulk', 'b0'))?.size).toBe(1);

    expect(s.listNodesByTag('special').length).toBe(5);
    expect(s.outboundEdges({ type: 'hub', id: 'h' }).length).toBe(10);
    expect(s.inboundEdges({ type: 'bulk', id: 'b0' }).length).toBe(1);
    await s.close();
  });

  it('stay consistent across replace, delete, edge delete, and cascade', async () => {
    const s = await open(join(dir, 'consistency.akg'));
    const state = internalState(s);

    // Tag replace: old leaves, new enters.
    s.putNode('doc', 'n1', { title: 't' }, ['old']);
    s.putNode('doc', 'n1', { title: 't' }, ['new']);
    expect(state.tagIndex.has('old')).toBe(false);
    expect(state.tagIndex.get('new')?.size).toBe(1);

    // Edge add/delete keeps both directions consistent.
    s.putNode('doc', 'n2', { title: 't' }, []);
    s.putEdge({ type: 'doc', id: 'n1' }, 'links', { type: 'doc', id: 'n2' }, {});
    expect(state.outIndex.get(nk('doc', 'n1'))?.size).toBe(1);
    expect(state.inIndex.get(nk('doc', 'n2'))?.size).toBe(1);
    s.deleteEdge({ type: 'doc', id: 'n1' }, 'links', { type: 'doc', id: 'n2' });
    expect(state.outIndex.has(nk('doc', 'n1'))).toBe(false);
    expect(state.inIndex.has(nk('doc', 'n2'))).toBe(false);

    // Node delete clears its tag entry.
    s.deleteNode('doc', 'n1');
    expect(state.tagIndex.has('new')).toBe(false);

    // Cascade clears incident edges (incl. a self-referential pair) and the node.
    s.putNode('doc', 'a', { title: 't' }, ['k']);
    s.putEdge({ type: 'doc', id: 'a' }, 'links', { type: 'doc', id: 'n2' }, {});
    s.putEdge({ type: 'doc', id: 'n2' }, 'links', { type: 'doc', id: 'a' }, {});
    const res = s.deleteNodeCascade('doc', 'a');
    expect(res).toEqual({ deletedInboundEdges: 1, deletedOutboundEdges: 1, deletedNode: true });
    expect(state.outIndex.has(nk('doc', 'a'))).toBe(false);
    expect(state.inIndex.has(nk('doc', 'a'))).toBe(false);
    expect(state.tagIndex.has('k')).toBe(false);
    await s.close();
  });

  it('rebuilds indexes from the committed WAL on reopen, honoring tag replace', async () => {
    const path = join(dir, 'replay.akg');
    const s = await open(path);
    // Two PUT_NODE records for the same node with different tags, both committed to
    // the WAL (no compaction) so reopen exercises the replay path, not hydrate.
    s.putNode('doc', 'r', { title: 't' }, ['first']);
    await s.commit();
    s.putNode('doc', 'r', { title: 't' }, ['second']);
    await s.close();

    const re = await open(path);
    const state = internalState(re);
    // The stale 'first' tag must not survive replay; only 'second' is indexed.
    expect(state.tagIndex.has('first')).toBe(false);
    expect(state.tagIndex.get('second')?.size).toBe(1);
    expect(re.listNodesByTag('first').length).toBe(0);
    expect(re.listNodesByTag('second').map((n) => n.id)).toEqual(['r']);
    await re.close();
  });
});

File: sdk/akg-go/indexes_test.go
package akg

import (
	"fmt"
	"path/filepath"
	"testing"
)

// TestSecondaryIndexesAreMatchSized proves PERF-1: ListNodesByTag, OutboundEdges,
// and InboundEdges draw from secondary indexes sized to the number of matches, not
// the total store size. A full O(total) scan would have to examine every node/edge;
// here the backing index sets hold exactly the matching entries regardless of how
// large the rest of the store is.
func TestSecondaryIndexesAreMatchSized(t *testing.T) {
	st, err := Open(filepath.Join(t.TempDir(), "idx.akg"))
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	defer st.Close()

	// A mass of untagged "bulk" nodes the indexes must NOT have to scan. Kept
	// under the 1000-entry WAL auto-flush threshold so the test exercises the
	// indexes, not repeated whole-file rewrites (that cost is IO-1's concern).
	const bulk = 500
	for i := 0; i < bulk; i++ {
		if _, err := st.PutNode("bulk", fmt.Sprintf("b%d", i), NodeFields{Title: "b"}, nil); err != nil {
			t.Fatalf("PutNode bulk: %v", err)
		}
	}
	// Five tagged nodes.
	for i := 0; i < 5; i++ {
		if _, err := st.PutNode("doc", fmt.Sprintf("d%d", i), NodeFields{Title: "d"}, []string{"special"}); err != nil {
			t.Fatalf("PutNode tagged: %v", err)
		}
	}
	// A hub with outbound edges to 10 of the bulk nodes, and one leaf with a single
	// inbound edge.
	if _, err := st.PutNode("hub", "h", NodeFields{Title: "hub"}, nil); err != nil {
		t.Fatalf("PutNode hub: %v", err)
	}
	for i := 0; i < 10; i++ {
		if err := st.PutEdge(NodeRef{Type: "hub", ID: "h"}, "links", NodeRef{Type: "bulk", ID: fmt.Sprintf("b%d", i)}, EdgeFields{}); err != nil {
			t.Fatalf("PutEdge: %v", err)
		}
	}

	// Index sets are sized to matches, independent of the 2000-node bulk.
	if got := len(st.state.tagIndex["special"]); got != 5 {
		t.Fatalf("tagIndex[special] size = %d, want 5 (must not include %d bulk nodes)", got, bulk)
	}
	hub := nodeIdentity{typeName: "hub", id: "h"}
	if got := len(st.state.outIndex[hub]); got != 10 {
		t.Fatalf("outIndex[hub] size = %d, want 10", got)
	}
	leaf := nodeIdentity{typeName: "bulk", id: "b0"}
	if got := len(st.state.inIndex[leaf]); got != 1 {
		t.Fatalf("inIndex[b0] size = %d, want 1", got)
	}

	// And the queries return exactly those matches.
	tagged, err := st.ListNodesByTag("special")
	if err != nil || len(tagged) != 5 {
		t.Fatalf("ListNodesByTag = %d nodes (err=%v), want 5", len(tagged), err)
	}
	out, err := st.OutboundEdges(NodeRef{Type: "hub", ID: "h"}, "")
	if err != nil || len(out) != 10 {
		t.Fatalf("OutboundEdges = %d (err=%v), want 10", len(out), err)
	}
	in, err := st.InboundEdges(NodeRef{Type: "bulk", ID: "b0"}, "")
	if err != nil || len(in) != 1 {
		t.Fatalf("InboundEdges = %d (err=%v), want 1", len(in), err)
	}
}

// TestSecondaryIndexConsistency exercises every mutation path that must keep the
// indexes in sync: tag replace, node delete, edge delete, cascade delete, and a
// reopen that rebuilds the indexes from persisted records.
func TestSecondaryIndexConsistency(t *testing.T) {
	path := filepath.Join(t.TempDir(), "consistency.akg")
	st, err := Open(path)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}

	// Replace tags: "old" must leave the index, "new" must enter it.
	if _, err := st.PutNode("doc", "n1", NodeFields{Title: "t"}, []string{"old"}); err != nil {
		t.Fatalf("PutNode: %v", err)
	}
	if _, err := st.PutNode("doc", "n1", NodeFields{Title: "t"}, []string{"new"}); err != nil {
		t.Fatalf("PutNode replace: %v", err)
	}
	if _, ok := st.state.tagIndex["old"]; ok {
		t.Fatal("tagIndex still has 'old' after tag replace")
	}
	if got := len(st.state.tagIndex["new"]); got != 1 {
		t.Fatalf("tagIndex[new] = %d, want 1", got)
	}

	// Edge add/delete keeps both directions consistent.
	if _, err := st.PutNode("doc", "n2", NodeFields{Title: "t"}, nil); err != nil {
		t.Fatalf("PutNode n2: %v", err)
	}
	if err := st.PutEdge(NodeRef{Type: "doc", ID: "n1"}, "links", NodeRef{Type: "doc", ID: "n2"}, EdgeFields{}); err != nil {
		t.Fatalf("PutEdge: %v", err)
	}
	n1 := nodeIdentity{typeName: "doc", id: "n1"}
	n2 := nodeIdentity{typeName: "doc", id: "n2"}
	if len(st.state.outIndex[n1]) != 1 || len(st.state.inIndex[n2]) != 1 {
		t.Fatal("edge indexes not populated after PutEdge")
	}
	if err := st.DeleteEdge(NodeRef{Type: "doc", ID: "n1"}, "links", NodeRef{Type: "doc", ID: "n2"}); err != nil {
		t.Fatalf("DeleteEdge: %v", err)
	}
	if len(st.state.outIndex[n1]) != 0 || len(st.state.inIndex[n2]) != 0 {
		t.Fatal("edge indexes not cleared after DeleteEdge")
	}

	// Delete node clears its tag index entry.
	if err := st.DeleteNode("doc", "n1"); err != nil {
		t.Fatalf("DeleteNode: %v", err)
	}
	if len(st.state.tagIndex["new"]) != 0 {
		t.Fatal("tagIndex[new] not cleared after DeleteNode")
	}

	// Cascade delete removes edges and node, leaving indexes empty.
	if _, err := st.PutNode("doc", "a", NodeFields{Title: "t"}, []string{"k"}); err != nil {
		t.Fatalf("PutNode a: %v", err)
	}
	if err := st.PutEdge(NodeRef{Type: "doc", ID: "a"}, "links", NodeRef{Type: "doc", ID: "n2"}, EdgeFields{}); err != nil {
		t.Fatalf("PutEdge a->n2: %v", err)
	}
	if err := st.PutEdge(NodeRef{Type: "doc", ID: "n2"}, "links", NodeRef{Type: "doc", ID: "a"}, EdgeFields{}); err != nil {
		t.Fatalf("PutEdge n2->a: %v", err)
	}
	res, err := st.DeleteNodeCascade("doc", "a")
	if err != nil {
		t.Fatalf("DeleteNodeCascade: %v", err)
	}
	if res.DeletedOutboundEdges != 1 || res.DeletedInboundEdges != 1 || !res.DeletedNode {
		t.Fatalf("cascade result = %+v, want 1 out / 1 in / deleted", res)
	}
	a := nodeIdentity{typeName: "doc", id: "a"}
	if len(st.state.outIndex[a]) != 0 || len(st.state.inIndex[a]) != 0 || len(st.state.tagIndex["k"]) != 0 {
		t.Fatal("indexes not cleared after cascade delete")
	}

	// Reopen: indexes must be rebuilt from persisted records so queries still work.
	if _, err := st.PutNode("doc", "survivor", NodeFields{Title: "t"}, []string{"keep"}); err != nil {
		t.Fatalf("PutNode survivor: %v", err)
	}
	if err := st.Close(); err != nil {
		t.Fatalf("Close: %v", err)
	}
	re, err := Open(path)
	if err != nil {
		t.Fatalf("reopen: %v", err)
	}
	defer re.Close()
	if got := len(re.state.tagIndex["keep"]); got != 1 {
		t.Fatalf("after reopen, tagIndex[keep] = %d, want 1 (index not rebuilt at load)", got)
	}
	survivors, err := re.ListNodesByTag("keep")
	if err != nil || len(survivors) != 1 {
		t.Fatalf("after reopen, ListNodesByTag(keep) = %d (err=%v), want 1", len(survivors), err)
	}
}

// TestSecondaryIndexReplayHonorsTagReplace covers the subtle case where committed
// WAL replay sees two PUT_NODE records for the same node with different tags: the
// rebuilt index must reflect only the latest tags, with no stale entry.
func TestSecondaryIndexReplayHonorsTagReplace(t *testing.T) {
	path := filepath.Join(t.TempDir(), "replay-replace.akg")
	st, err := Open(path)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	// Two PUT_NODE records for the same node, both committed to the WAL (no
	// compaction) so reopen exercises replay, not hydrate.
	if _, err := st.PutNode("doc", "r", NodeFields{Title: "t"}, []string{"first"}); err != nil {
		t.Fatalf("PutNode first: %v", err)
	}
	if err := st.Commit(); err != nil {
		t.Fatalf("Commit: %v", err)
	}
	if _, err := st.PutNode("doc", "r", NodeFields{Title: "t"}, []string{"second"}); err != nil {
		t.Fatalf("PutNode second: %v", err)
	}
	if err := st.Close(); err != nil {
		t.Fatalf("Close: %v", err)
	}

	re, err := Open(path)
	if err != nil {
		t.Fatalf("reopen: %v", err)
	}
	defer re.Close()
	if _, ok := re.state.tagIndex["first"]; ok {
		t.Fatal("stale 'first' tag survived WAL replay")
	}
	if got := len(re.state.tagIndex["second"]); got != 1 {
		t.Fatalf("tagIndex[second] = %d, want 1", got)
	}
	if got, _ := re.ListNodesByTag("first"); len(got) != 0 {
		t.Fatalf("ListNodesByTag(first) = %d, want 0", len(got))
	}
	if got, _ := re.ListNodesByTag("second"); len(got) != 1 {
		t.Fatalf("ListNodesByTag(second) = %d, want 1", len(got))
	}
}
